pkg/gatewaysecret/handler_test.go

```diff
@@ -0,0 +1,356 @@
+					},
+				},
+				caCert: certmanagerv1.Certificate{
+					Status: certmanagerv1.CertificateStatus{
+						NotBefore: &apimetav1.Time{
+							Time: time.Date(2024, 11, 1, 0, 0, 5, 0, time.UTC),
+						},
+					},
+				},
+			},
+			want: true,
+		},
+	}
+	for _, testcase := range tests {
+		t.Run(testcase.name, func(t *testing.T) {
+			if got := gatewaysecret.GatewaySecretRequiresUpdate(
+				testcase.args.gwSecret, testcase.args.caCert); got != testcase.want {
+				t.Errorf("GatewaySecretRequiresUpdate() = %v, want %v", got, testcase.want)
+			}
+		})
+	}
+}
+
+func TestCopyRootSecretDataIntoGatewaySecret(t *testing.T) {
+	t.Parallel()
+
+	// Current gateway secret
+	gwSecret := &apicorev1.Secret{
+		Data: map[string][]byte{
+			"tls.crt": []byte(oldTLSCertValue),
+			"tls.key": []byte(oldTLSKeyValue),
+			"ca.crt":  []byte(oldCACertValue),
+		},
+	}
+
+	rootSecret := &apicorev1.Secret{
+		Data: map[string][]byte{
+			"tls.crt": []byte(newTLSCertValue),
+			"tls.key": []byte(newTLSKeyValue),
+			"ca.crt":  []byte(newCACertValue),
+		},
+	}
+
+	gatewaysecret.CopyRootSecretDataIntoGatewaySecret(gwSecret, rootSecret)
+
+	require.Equal(t, newTLSCertValue, string(gwSecret.Data["tls.crt"]))
+	require.Equal(t, newTLSKeyValue, string(gwSecret.Data["tls.key"]))
+	require.Equal(t, newCACertValue, string(gwSecret.Data["ca.crt"]))
+}
+
+type MockSecretManager struct {
+	findGatewaySecretFunc    func(ctx context.Context) (*apicorev1.Secret, error)
+	getRootCACertificateFunc func(ctx context.Context) (certmanagerv1.Certificate, error)
+	createFunc               func(ctx context.Context, secret *apicorev1.Secret) error
+	updateFunc               func(ctx context.Context, secret *apicorev1.Secret) error
+}
+
+func (m MockSecretManager) FindGatewaySecret(ctx context.Context) (*apicorev1.Secret, error) {
+	return m.findGatewaySecretFunc(ctx)
+}
+
+func (m MockSecretManager) GetRootCACertificate(ctx context.Context) (certmanagerv1.Certificate, error) {
+	return m.getRootCACertificateFunc(ctx)
+}
+
+func (m MockSecretManager) Create(ctx context.Context, secret *apicorev1.Secret) error {
+	return m.createFunc(ctx, secret)
+}
+
+func (m MockSecretManager) Update(ctx context.Context, secret *apicorev1.Secret) error {
+	return m.updateFunc(ctx, secret)
+}
+
+func TestWatchEventsNewGatewaySecret(t *testing.T) {
+	t.Parallel()
+
+	findGatewaySecretFunc := func(ctx context.Context) (*apicorev1.Secret, error) {
+		return nil, &meta.NoResourceMatchError{}
+	}
+	createFunc := func(ctx context.Context, gwSecret *apicorev1.Secret) error {
+		require.Equal(t, newTLSCertValue, string(gwSecret.Data["tls.crt"]))
+		require.Equal(t, newTLSKeyValue, string(gwSecret.Data["tls.key"]))
+		require.Equal(t, newCACertValue, string(gwSecret.Data["ca.crt"]))
+
+		return nil
+	}
+	mockManager := MockSecretManager{findGatewaySecretFunc: findGatewaySecretFunc, createFunc: createFunc}
+	handler := gatewaysecret.GatewaySecretHandler{SecretManager: mockManager}
```

Comment:
I think this test is too complex to set up; it indicates a "too much coupling" condition.
Take a look at the event loop. You don't need entire `*GatewaySecretHandler` instance there. You use just one function. The event loop test should just test if this function is called with correct arguments for certain event types, and that the error is correrctly reported.
In that case you only need to mock one function - the one that is invoked in the event loop.